src/drivers/amd/rembrandt/acp_dma.c

```diff
@@ -0,0 +1,662 @@
+static int acp_dma_set_config(struct dma_chan_data *channel,
+			   struct dma_sg_config *config)
+{
+	uint32_t dir;
+
+	channel->direction = config->direction;
+	dir = config->direction;
+	if (config->cyclic) {
+		tr_err(&acpdma_tr,
+			"DMA: cyclic configurations are not supported");
+		return -EINVAL;
+	}
+	if (config->scatter) {
+		tr_err(&acpdma_tr,
+			"DMA: scatter is not supported Chan.Index %d scatter %d",
+			channel->index, config->scatter);
+		return -EINVAL;
+	}
+	return dma_setup(channel, &config->elem_array, dir);
+}
+
+static int acp_dma_probe(struct dma *dma)
+{
+	struct acp_dma_chan_data *acp_dma_chan;
+	int channel;
+
+	if (dma->chan) {
+		tr_err(&acpdma_tr, "DMA: Already probe");
+		return -EEXIST;
+	}
+	dma->chan = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM,
+			    dma->plat_data.channels *
+			    sizeof(struct dma_chan_data));
+	if (!dma->chan) {
+		tr_err(&acpdma_tr, "DMA: unable to allocate channel context");
+		return -ENOMEM;
+	}
+	for (channel = 0; channel < dma->plat_data.channels; channel++) {
+		dma->chan[channel].dma = dma;
+		dma->chan[channel].index = channel;
+		dma->chan[channel].status = COMP_STATE_INIT;
+		acp_dma_chan = rzalloc(SOF_MEM_ZONE_SYS_RUNTIME, 0,
+				   SOF_MEM_CAPS_RAM,
+				   sizeof(struct acp_dma_chan_data));
+		if (!acp_dma_chan) {
+			rfree(dma->chan);
+			tr_err(&acpdma_tr, "acp-dma: %d channel %d private data alloc failed",
+			       dma->plat_data.id, channel);
+			return -ENOMEM;
+		}
+		dma_chan_set_data(&dma->chan[channel], acp_dma_chan);
+	}
+	return 0;
+}
+
+static int acp_dma_remove(struct dma *dma)
+{
+	int channel;
+
+	if (!dma->chan) {
+		tr_err(&acpdma_tr, "DMA: Invalid remove call");
+		return 0;
+	}
+	for (channel = 0; channel < dma->plat_data.channels; channel++)
+		rfree(dma->chan[channel].priv_data);
+	rfree(dma->chan);
+	dma->chan = NULL;
+
+	return 0;
+}
+
+static int acp_dma_interrupt(struct dma_chan_data *channel, enum dma_irq_cmd cmd)
+{
+	uint32_t status;
+
+	if (channel->status == COMP_STATE_INIT)
+		return 0;
+
+	switch (cmd) {
+	case DMA_IRQ_STATUS_GET:
+		status =  dma_reg_read(channel->dma, ACP_DSP0_INTR_STAT) & 0xFF;
+		return status & (1 << channel->index);
+	case DMA_IRQ_CLEAR:
+		status =  dma_reg_read(channel->dma, ACP_DSP0_INTR_STAT);
+		status  = status & (1 << channel->index);
+		dma_reg_write(channel->dma, ACP_DSP0_INTR_STAT, status);
+		return 0;
+	case DMA_IRQ_MASK:
+		status =  dma_reg_read(channel->dma, ACP_DSP0_INTR_CNTL);
+		status  = status  & (~(1 << channel->index));
+		dma_reg_write(channel->dma, ACP_DSP0_INTR_CNTL, status);
+		return 0;
+	case DMA_IRQ_UNMASK:
+		status =  dma_reg_read(channel->dma, ACP_DSP0_INTR_CNTL);
+		status  = status | (1 << channel->index);
+		dma_reg_write(channel->dma, ACP_DSP0_INTR_CNTL, status);
+		return 0;
+	default:
+		return -EINVAL;
+	}
+}
+
+static int acp_dma_get_attribute(struct dma *dma, uint32_t type, uint32_t *value)
+{
+	switch (type) {
```

Comment:
if im reading this right, a lot of these functions are the same, you should make this a helper you just install, so you don't have redefine the same thing all over the place

Reply:
will try to reuse the code in up coming platforms.